Dynamic tabs: use buttons rather than links

patrickhlauke opened the pull request:
- change docs
- tweak styles to neutralise border and background

this is more semantically correct. note that currently, the forced `:focus` in reboot styles is making this look rather ugly - see separate PR https://github.com/twbs/bootstrap/pull/32631 to remove that cargo culted (?) style from reboot.

Preview: https://deploy-preview-32630--twbs-bootstrap.netlify.app/docs/5.0/components/navs-tabs/#javascript-behavior

patrickhlauke committed:
Dynamic tabs: use buttons rather than links

- change docs
- tweak styles to neutralise border and background

patrickhlauke commented:
set to draft for now while we wait for https://github.com/twbs/bootstrap/pull/32631 - once that is in place (removing the forced `:focus` style on buttons), this works cleanly.

only minor concerns:

- `.nav-link` used on buttons may not seem logical, but changing the class name now would be a breaking change
- the buttons have black, rather than link-blue, text

patrickhlauke committed:
Merge branch 'main' into patrickhlauke-navs-tabs

patrickhlauke committed:
Merge branch 'main' into patrickhlauke-navs-tabs

mdo reviewed (commented):
We should include some docs mentions here for why we use `<button>` elements here.

patrickhlauke committed:
Merge branch 'main' into patrickhlauke-navs-tabs

patrickhlauke committed:
Add mention that tabs should be <button> elements

per @mdo's suggestion

patrickhlauke committed:
Update js unit and visual test accordingly

- replace links with buttons
- make one specific test that uses links instead of buttons, as we still want to support it despite it being non-semantically appropriate
- remove tests involving dropdown tabs, as they're discouraged (though still silently supported)

patrickhlauke committed:
Tweak tests

patrickhlauke committed:
Fix tests

patrickhlauke committed:
Tests

patrickhlauke committed:
Merge branch 'main' into patrickhlauke-navs-tabs

patrickhlauke committed:
Remove test for unusued and frankly inappropriate structure/handling

patrickhlauke commented:
> We should include some docs mentions here for why we use `<button>` elements here.

added a short sentence to explain. also took opportunity to update the visual and unit tests

patrickhlauke committed:
Tweak code example in events part of the docs

XhmikosR commented on scss/_nav.scss:
```diff
@@ -80,6 +81,8 @@
 
 .nav-pills {
   .nav-link {
+    background: none;
```
Should the shorthand be used here and the other CSS change?

Also, not sure about `border: 0` / `border: none`

patrickhlauke replied:
we do use the `background` shorthand in a few places (to set things, not to neutralise them, admittedly).

ditto for `border: 0`. we use that in various places. no `border: none`

ffoodd commented:
You may use `isolation: isolate` on the focus state to ensure the focused item appears above the active one:

**without** `isolation` |  **with** `isolation
![Capture d’écran de 2021-01-07 10-58-30](https://user-images.githubusercontent.com/2890570/103879111-b2082100-50d7-11eb-8b9d-07c7bfd1b81a.png) | 
![Capture d’écran de 2021-01-07 10-58-17](https://user-images.githubusercontent.com/2890570/103879119-b46a7b00-50d7-11eb-8f44-06e292eacb59.png)

Simple improvement, might be used elsewhere since [support is very well aligned with our browsers list](https://caniuse.com/mdn-css_properties_isolation).

patrickhlauke committed:
Merge branch 'main' into patrickhlauke-navs-tabs

patrickhlauke commented:
> You may use `isolation: isolate` on the focus state to ensure the focused item appears above the active one

on which element/class are you adding it to? i just messed around in Firefox devtools, and seems adding it to `.nav-link` and to `.nav-item`, but no joy...

ffoodd commented:
On `.nav-link`'s hover/focus state, since they're the one holding borders.

patrickhlauke committed:
Merge branch 'main' into patrickhlauke-navs-tabs

patrickhlauke committed:
Add isolation:isolate to prevent focus being overlapped

https://github.com/twbs/bootstrap/pull/32630#issuecomment-756015766

patrickhlauke committed:
Merge branch 'patrickhlauke-navs-tabs' of https://github.com/twbs/bootstrap into patrickhlauke-navs-tabs

patrickhlauke commented:
@ffoodd good call on the isolation thing. done.

patrickhlauke committed:
Merge branch 'main' into patrickhlauke-navs-tabs

patrickhlauke committed:
Merge branch 'main' into patrickhlauke-navs-tabs

patrickhlauke commented:
@mdo good to merge?

patrickhlauke committed:
Merge branch 'main' into patrickhlauke-navs-tabs

patrickhlauke commented:
As the carousel buttons stuff has been merged, would be keen to see this make its way into main as well. Any concerns on this?

patrickhlauke committed:
Merge branch 'main' into patrickhlauke-navs-tabs

mdo reviewed (approved):
LGTM. I'm fine including in Beta 2 or holding for Beta 3.

XhmikosR commented:
1. Do we have a test with JS test `a` tags?
2. I think the wording needs some tweaking:

  >Dynamic tabbed interfaces, as described in the WAI ARIA Authoring Practices, require role="tablist", role="tab", role="tabpanel", and additional aria- attributes in order to convey their structure, functionality and current state to users of assistive technologies (such as screen readers). As a best practice, we recommend using <button> elements for the tabs, as these are controls that trigger a dynamic change, rather than links that navigate to a new page or location.

  I feel like It's not clear that if someone is using `button`s, the `role` isn't needed.

patrickhlauke commented:
> > Dynamic tabbed interfaces, as described in the [WAI ARIA Authoring Practices](https://www.w3.org/TR/wai-aria-practices/#tabpanel), require `role="tablist"`, `role="tab"`, `role="tabpanel"`, and additional `aria-` attributes in order to convey their structure, functionality and current state to users of assistive technologies (such as screen readers). As a best practice, we recommend using `<button>` elements for the tabs, as these are controls that trigger a dynamic change, rather than links that navigate to a new page or location.
> 
> I feel like It's not clear that if someone is using `button`s, the `role` isn't needed.

`role="tab"` *is* needed even when using `<button>`

patrickhlauke commented:
> LGTM. I'm fine including in Beta 2 or holding for Beta 3.

ah darn, sorry...forgot about the whole beta 2 / beta 3 aspect...got confused with the release of 4.6.0 recently thinking "so why is this still hanging there and not being merged?"

ok, i'm happy either way, whether it makes it into 2 or 3

patrickhlauke commented:
> 1. Do we have a test with JS test `a` tags?

https://github.com/twbs/bootstrap/pull/32630/files#diff-e808aa1f3605a07d2a11a7dcb2f81c71f6bd2fd9257979ed460a3f3141d07bc8R48

unless you meant replicating every test to run both for `<button>` and `<a>` elements?

XhmikosR commented:
> 
> 
> > 1. Do we have a test with JS test `a` tags?
> 
> https://github.com/twbs/bootstrap/pull/32630/files#diff-e808aa1f3605a07d2a11a7dcb2f81c71f6bd2fd9257979ed460a3f3141d07bc8R48
> 
> unless you meant replicating every test to run both for `<button>` and `<a>` elements?

Not all of them, just one test like you did with that other PR of yours.

> `role="tab"` _is_ needed even when using `<button>`

Sorry, missed that due to horizontal scrolling :)

patrickhlauke commented:
> > > 1. Do we have a test with JS test `a` tags?
> > 
> > 
> > https://github.com/twbs/bootstrap/pull/32630/files#diff-e808aa1f3605a07d2a11a7dcb2f81c71f6bd2fd9257979ed460a3f3141d07bc8R48
> > unless you meant replicating every test to run both for `<button>` and `<a>` elements?
> 
> Not all of them, just one test like you did with that other PR of yours.

So we're good here, as I did it for this PR as well as per link above :)

XhmikosR commented:
Cool. One last thing. I wouldn't remove any tests. Even if something is inappropriate, if it's still supported, it should be tested.

patrickhlauke commented:
> Cool. One last thing. I wouldn't remove any tests. Even if something is inappropriate, if it's still supported, it should be tested.

assuming it's about the dynamic tabs using dropdown menus, as that pattern is completely wrong/inappropriate (from an accessibility and usability pov), not sure it makes sense to keep it in. in fact, tempted for v5 to strip out any code that handles this (as we already warned about/deprecated it in v4)

patrickhlauke commented on js/tests/unit/tab.spec.js:
```diff
@@ -277,55 +301,6 @@ describe('Tab', () => {
 
       firstTab.show()
     })
-
```
this test seems to test something that was never documented as a kosher technique/structure? it's invalid markup (having buttons inside of links) and should never be used, and i don't think we show this in any docs/examples (checked old v2 and v3 docs too, and couldn't see any example of this). feels safe to remove?

rohit2sharma95 replied:
This one?
**_should not fire hidden when hide is prevented_**

patrickhlauke replied:
yeah but not using close buttons this way. maybe programmatically removed ones?

patrickhlauke replied:
i've put the code for this and the dropdown tabs back for now...will need some more careful work here (to reformulate this one into a programmatic handling of tabs disappearing/being removed, and the other one a "breaking" change that should happen when we remove the actual logic in tab.js that handles this as well)

patrickhlauke committed:
Put back 2 of the removed tests

Leaving these here for now. The test for removed tabs should be redone so that tabs are removed programmatically (as the approach of having that close button inside the link is invalid and broken markup). The test for dropdowns should be removed together we actually ripping out the handling for dropdowns in the tab.js code (arguably a breaking change, though we discouraged this for a few versions and effectively "deprecated" it)

patrickhlauke commented on js/tests/visual/tab.html:
```diff
@@ -159,47 +159,14 @@ <h4>Tabs with fade (no initially active pane)</h4>
         </div>
       </div>
 
-      <h4>Tabs with nav (with fade)</h4>
```
Removed this as it's a weirdly invalid example, markup-wise. Not sure what this was supposed to show/visually test for...

XhmikosR replied:
Maybe just nav without ul?

patrickhlauke replied:
right, fixed up (using the suggested pattern from the example using `<nav>` in https://getbootstrap.com/docs/5.0/components/navs-tabs/#javascript-behavior, and removing the dropdown) and reintroduced the visual test. this is now good to go from my end i'd say

patrickhlauke commented:
Put back the two tests that I had originally shelved. This should keep this PR focused and non-breaking

patrickhlauke committed:
Merge branch 'main' into patrickhlauke-navs-tabs

patrickhlauke committed:
Reintroduce and fix visual test

patrickhlauke committed:
Merge branch 'main' into patrickhlauke-navs-tabs

patrickhlauke commented:
:rocket:

MickL commented:
Has this been published with beta2 6 days ago? For me buttons still dont look good inside `nav-pills`:

<img width="294" alt="Bildschirmfoto 2021-02-16 um 13 56 01" src="https://user-images.githubusercontent.com/5103210/108065912-f257a900-705e-11eb-80e4-e096557430f4.png">

Using the exact code from the docs:

```
<ul class="nav nav-pills">
  <li class="nav-item">
    <button class="nav-link active" aria-current="page">Active</button>
  </li>
  <li class="nav-item">
    <button class="nav-link">Link</button>
  </li>
  <li class="nav-item">
    <button class="nav-link">Link</button>
  </li>
  <li class="nav-item">
    <button class="nav-link disabled" tabindex="-1" aria-disabled="true">Disabled</button>
  </li>
</ul>
```

patrickhlauke commented:
@MickL if you're using the exact same code as in the docs, but it's looking different from https://getbootstrap.com/docs/5.0/components/navs-tabs/#javascript-behavior, then it's likely some other modification/customisation you're doing that's affecting this?

MickL commented:
@patrickhlauke obviously it is exactly the same BUT i changed `<a>` for `<button>`

ffoodd commented:
@MickL Our current docs example (linked by Patrick) is working fine with `button`, this is why this PR got merged and your issue closed.

If you still see any issue related to this in beta 21, please open a new issue—and stop spamming an already merged PR that we cannot re-open.
